In [3]:
import numpy as np
from mp_api.client import MPRester
from pymatgen.core.operations import SymmOp
from pymatgen.symmetry.analyzer import SpacegroupAnalyzer
from pymatgen.io.cif import CifWriter
from jupyter_jsmol.pymatgen import quick_view


In [4]:
mp_key = "jBtrGioeefhRZ0bpN5xCDp8bQrH6O0HK"
mp_id = "mp-30459"

## Téléchargement fichier CIF

In [8]:
with MPRester(mp_key) as m:
    cif_data = m.get_structure_by_material_id(mp_id)

CifWriter(cif_data).write_file("mp-30459.cif")
print("Succès du téléchargement du fichier sous le nom :mp-30459.cif ")

Retrieving MaterialsDoc documents:   0%|          | 0/1 [00:00<?, ?it/s]

Succès du téléchargement du fichier sous le nom :mp-30459.cif 


## Obtention informations générale sur les opérations de symmétries

In [9]:
conv_struc = SpacegroupAnalyzer(cif_data).get_conventional_standard_structure()
symmops = SpacegroupAnalyzer(conv_struc).get_symmetry_operations()


#print("Affichage de l'ensemble des opérations de symmétries uniques au sein du matériau")
#for i, symmop in enumerate(symmops, start=1):
    
    #print(f"Operation {i}:")
    #print("Tenseur de rotation:")
    #print(symmop.rotation_matrix)
    #print("Vecteur de translation:")
    #print(symmop.translation_vector)

print("Decommenter les lignes afin d'obtenir toutes les opérations de symmetries uniques, ceci a été commenté dans le but de réduire la charge visuelle sur l'affichage github.Ces lignes de codes ont été effectuées pour aider à la visualisation des différentes opérations et afin de faciliter le choix dans la suite.")


Decommenter les lignes afin d'obtenir toutes les opérations de symmetries uniques, ceci a été commenté dans le but de réduire la charge visuelle sur l'affichage github.Ces lignes de codes ont été effectuées pour aider à la visualisation des différentes opérations et afin de faciliter le choix dans la suite.


## Operateur de symmetrie 1 (Opération 3) : 

In [10]:
view = quick_view(cif_data,"packed",conventional = "True")
display(view)
view.script('draw SYMOP 2 {atomno=1}')
print(f"Operation {2}:")
print("Tenseur de rotation:")
print(symmops[1].rotation_matrix)
print("Vecteur de translation:")
print(symmops[1].translation_vector)


rotation_angle = np.degrees(np.arccos((np.trace(symmops[1].rotation_matrix) - 1) / 2))

print("Angle de rotation")
print(rotation_angle)
print("Ordre Opération = ", round(360/rotation_angle))

#calcul positions initial et finale de l'atome étudié
pos_init = conv_struc.sites[1 -1].frac_coords
print("Position initiale de l'atome 1 :",pos_init)
pos_final = symmops[1].operate(pos_init)
print("Positifion finale de l'atome 1 :",pos_final)



JsmolView(layout=Layout(align_self='stretch', height='400px'))

Operation 2:
Tenseur de rotation:
[[ 0.  1.  0.]
 [-1.  0.  0.]
 [ 0.  0. -1.]]
Vecteur de translation:
[0. 0. 0.]
Angle de rotation
180.0
Ordre Opération =  2
Position initiale de l'atome 1 : [0.5 0.  0. ]
Positifion finale de l'atome 1 : [ 0.  -0.5  0. ]


Cette opération de symétrie implique une rotation de 180 degrés autour de l'axe z sans translation. L'effet de cette opération sur l'atome 1 est de déplacer sa position initiale [0.5 0. 0.] vers une nouvelle position [ 0. -0.5 0.] L'opération envoie l'atome de la position (x,y,z) dans la position (y,-x,-z).

$$
\text{Tenseur de rotation:} \quad
\begin{pmatrix}
0 & 1 & 0 \\
-1 & 0 & 0 \\
0 & 0 & -1 \\
\end{pmatrix}
\cdot
\begin{pmatrix}
x \\
y \\
z \\
\end{pmatrix}
+
\text{Vecteur de translation:} \quad
\begin{pmatrix}
0 \\
0 \\
0 \\
\end{pmatrix}
=
\begin{pmatrix}
y \\
-x \\
-z \\
\end{pmatrix}
$$

## Element de symmetrie 2 (Operation 41) :

In [11]:
view = quick_view(cif_data,"packed",conventional = "True")
display(view)
view.script('draw SYMOP 41 {atomno=2}')
print(f"Operation {41}:")
print("Tenseur de rotation:")
print(symmops[40].rotation_matrix)
print("Vecteur de translation:")
print(symmops[40].translation_vector)


rotation_angle = np.degrees(np.arccos((np.trace(symmops[40].rotation_matrix) - 1) / 2))

print("Angle de rotation")
print(rotation_angle)
print("Ordre Opération = ", round(360/rotation_angle))

pos_init = conv_struc.sites[2 -1].frac_coords
print("Position initiale de l'atome 2 :",pos_init)
pos_final = symmops[40].operate(pos_init)
print("Positifion finale de l'atome 2 :",pos_final)



JsmolView(layout=Layout(align_self='stretch', height='400px'))

Operation 41:
Tenseur de rotation:
[[0. 1. 0.]
 [0. 0. 1.]
 [1. 0. 0.]]
Vecteur de translation:
[0.5 0.5 0. ]
Angle de rotation
120.00000000000001
Ordre Opération =  3
Position initiale de l'atome 2 : [0.5 0.5 0.5]
Positifion finale de l'atome 2 : [1.  1.  0.5]


Cette opération de symétrie implique une rotation de 120 degrés autour de l'axe z avec translation. L'effet de cette opération sur l'atome 2 est de déplacer sa position initiale [0.5 0.5 0.5] vers une nouvelle position [1. 1. 0.5].L'opération envoie  l'atome de la position (x,y,z) dans la position (y+1/2,z+1/2,x).

$$
\text{Tenseur de rotation:} \quad
\begin{pmatrix}
0 & 1 & 0 \\
0 & 0 & 1 \\
1 & 0 & 0 \\
\end{pmatrix}
\cdot
\begin{pmatrix}
x \\
y \\
z \\
\end{pmatrix}
+
\text{Vecteur de translation:} \quad
\begin{pmatrix}
0.5 \\
0.5 \\
0 \\
\end{pmatrix}
=
\begin{pmatrix}
y+1/2 \\
z+1/2 \\
x \\
\end{pmatrix}
$$

## Element de symmetrie 3 (Operation 56) :

In [13]:
view = quick_view(cif_data,"packed",conventional = "True")
display(view)
view.script('draw SYMOP 56 {atomno=3}')
print(f"Operation {56}:")
print("Tenseur de rotation:")
print(symmops[55].rotation_matrix)
print("Vecteur de translation:")
print(symmops[55].translation_vector)


rotation_angle = np.degrees(np.arccos((np.trace(symmops[55].rotation_matrix) - 1) / 2))

print("Angle de rotation")
print(rotation_angle)
print("Ordre Opération = ", round(360/rotation_angle))

pos_init = conv_struc.sites[3 -1].frac_coords
print("Position initiale de l'atome 3 :",pos_init)
pos_final = symmops[55].operate(pos_init)
print("Positifion finale de l'atome 3 :",pos_final)




JsmolView(layout=Layout(align_self='stretch', height='400px'))

Operation 56:
Tenseur de rotation:
[[ 0. -1.  0.]
 [-1.  0.  0.]
 [ 0.  0.  1.]]
Vecteur de translation:
[0.5 0.  0.5]
Angle de rotation
90.0
Ordre Opération =  4
Position initiale de l'atome 3 : [0.  0.  0.5]
Positifion finale de l'atome 3 : [0.5 0.  1. ]


Cette opération de symétrie implique une rotation de 90 degrés autour de l'axe y, suivie d'une translation de 0.5 unités le long de l'axe x et de 0.5 unités le long de l'axe z. L'effet de cette opération sur l'atome 3 est de déplacer sa position initiale [0. 0. 0.5] vers une nouvelle position [0.5 0. 1.]. L'opération envoie l'atome de la position (x,y,z) dans la position (-y+1/2,-x,z+1/2).

$$
\text{Tenseur de rotation:} \quad
\begin{pmatrix}
0 & -1 & 0 \\
-1 & 0 & 0 \\
0 & 0 & 1 \\
\end{pmatrix}
\cdot
\begin{pmatrix}
x \\
y \\
z \\
\end{pmatrix}
+
\text{Vecteur de translation:} \quad
\begin{pmatrix}
0.5 \\
0 \\
0.5 \\
\end{pmatrix}
=
\begin{pmatrix}
-y+1/2 \\
-x \\
z+1/2 \\
\end{pmatrix}
$$